### PR TITLE
feat: create poll

### PR DESCRIPTION
**Pull request by ice-hector**
### What does this PR do?
This PR contains a new add poll flow inside the post-creation flow

### Changes brought by this PR
Add a poll button in the toolbar
Poll creation UI (adding title and answers)
Remove the poll using the close button
Add answers
Change poll length modal
Poll field validation (also handled enabled/disabled send post button stated)

### Additional information
Screenshot 2024-10-14 at 19 10 12
Screenshot 2024-10-14 at 19 09 44
Screenshot 2024-10-14 at 19 09 47
Screenshot 2024-10-14 at 19 09 49
Screenshot 2024-10-14 at 19 09 53
Screenshot 2024-10-14 at 19 09 56
Screenshot 2024-10-14 at 19 10 08

### CODE REVIEW

File: lib/app/features/feed/views/components/text_editor/text_editor.dart [hunk 1]
```diff
@@ -1,15 +1,101 @@
...
-  const TextEditor(this.controller, {super.key, this.placeholder});
+  const TextEditor(
+    this.controller, {
+    super.key,
+    this.placeholder,
+  });
   final QuillController controller;
   final String? placeholder;
 
+  Future<void> _showPollLengthTimeModal(BuildContext context) async {
```
Comment (ice-tychon) [inline lib/app/features/feed/views/components/text_editor/text_editor.dart:39]: Should we extract these 2 new methods to a separate class/file?

File: lib/app/features/feed/views/components/text_editor/text_editor.dart [hunk 2]
```diff
@@ -1,15 +1,101 @@
...
+    for (var i = 0; i < delta.operations.length; i++) {
+      final operation = delta.operations[i];
+      final length = operation.length ?? 1;
+
+      if (operation.isInsert && operation.data is Map<String, dynamic>) {
+        final data = operation.data! as Map<String, dynamic>;
+
+        if (data.containsKey('custom')) {
+          final customData = data['custom'];
+          if (customData is Map<String, dynamic> && customData.containsKey(textEditorPollKey)) {
```
Comment (ice-tychon) [inline lib/app/features/feed/views/components/text_editor/text_editor.dart:62]: Code duplication here and on the 70th line

File: lib/app/features/feed/views/components/text_editor/components/toolbar_buttons/text_editor_poll_button/text_editor_poll_button.dart [hunk 3]
```diff
@@ -3,17 +3,39 @@
...
+        textEditorController.document.insert(
+          index,
+          BlockEmbed.custom(
+            TextEditorPollEmbed(),
+          ),
+        );
+
+        textEditorController.moveCursorToPosition(index + 1);
+
+        QuillEditorConfigurations(
```
Comment (ice-tychon) [inline lib/app/features/feed/views/components/text_editor/components/toolbar_buttons/text_editor_poll_button/text_editor_poll_button.dart:33]: What does this block do? It creates a configuration and does nothing with it.

File: lib/app/features/feed/create_post/views/pages/create_post_modal/create_post_modal.dart [hunk 4]
```diff
@@ -38,9 +45,75 @@ class CreatePostModal extends HookWidget {
   }
 
   @override
-  Widget build(BuildContext context) {
+  Widget build(BuildContext context, WidgetRef ref) {
     final textEditorController = useQuillController();
     final hasContent = useTextEditorHasContent(textEditorController);
+    final hasPoll = useState(false);
+
+    useEffect(
```
Comment (ice-tychon) [inline lib/app/features/feed/create_post/views/pages/create_post_modal/create_post_modal.dart:53]: Please, extract it to a standalone hook

File: lib/app/features/feed/create_post/views/pages/create_post_modal/create_post_modal.dart [hunk 5]
```diff
@@ -38,9 +45,75 @@ class CreatePostModal extends HookWidget {
...
+        void checkPoll() {
+          final delta = textEditorController.document.toDelta();
+          var pollExists = false;
+
+          for (final operation in delta.operations) {
+            if (operation.isInsert && operation.data is Map<String, dynamic>) {
+              final data = operation.data! as Map<String, dynamic>;
+              if (data.containsKey('custom')) {
+                final customData = data['custom'];
+                if (customData is Map<String, dynamic> &&
```
Comment (ice-tychon) [inline lib/app/features/feed/create_post/views/pages/create_post_modal/create_post_modal.dart:64]: Code duplication here and on the 72nd line

File: lib/app/features/feed/views/components/text_editor/components/custom_blocks/text_editor_poll_block/components/poll_answer_item/poll_answer_item.dart [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    final focusNode = useFocusNode();
+    final hasFocus = useNodeFocused(focusNode);
+
+    final answer = ref.watch(pollAnswersNotifierProvider)[index];
+    final textController = useTextEditingController(text: answer.text);
+
+    const characterLimit = 25;
+    final remainingCharacters = characterLimit - textController.text.length;
+
+    useEffect(
```
Comment (ice-tychon) [inline lib/app/features/feed/views/components/text_editor/components/custom_blocks/text_editor_poll_block/components/poll_answer_item/poll_answer_item.dart:31]: Use `useOnInit` to simplify this a little bit

File: lib/app/features/feed/views/components/text_editor/components/custom_blocks/text_editor_poll_block/text_editor_poll_block.dart [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+class TextEditorPollEmbed extends CustomBlockEmbed {
+  TextEditorPollEmbed() : super(textEditorPollKey, '');
+}
+
+///
+/// Embed builder for [TextEditorPollBuilder].
+///
+
+class TextEditorPollBuilder extends EmbedBuilder {
+  TextEditorPollBuilder({this.onPollLengthPress, this.onRemovePollPress});
```
Comment (ice-tychon) [inline lib/app/features/feed/views/components/text_editor/components/custom_blocks/text_editor_poll_block/text_editor_poll_block.dart:27]: nit: Formatting looks odd

File: lib/app/features/feed/create_post/views/pages/create_post_modal/create_post_modal.dart [hunk 8]
```diff
@@ -38,9 +45,75 @@ class CreatePostModal extends HookWidget {
...
+
+        return () {
+          textEditorController.removeListener(checkPoll);
+        };
+      },
+      [textEditorController],
+    );
+
+    bool isPollValid() {
+      final pollTitle = ref.watch(pollTitleNotifierProvider);
```
Comment (ice-alcides) [inline lib/app/features/feed/create_post/views/pages/create_post_modal/create_post_modal.dart:100]:
```
  final pollTitle = ref.watch(pollTitleNotifierProvider);
  final pollAnswers = ref.watch(pollAnswersNotifierProvider);
```

change to `ref.read`

File: lib/app/features/feed/create_post/views/pages/create_post_modal/create_post_modal.dart [hunk 9]
```diff
@@ -38,9 +45,75 @@ class CreatePostModal extends HookWidget {
...
+    final hasPoll = useState(false);
+
+    useEffect(
+      () {
+        void checkPoll() {
+          final delta = textEditorController.document.toDelta();
+          var pollExists = false;
+
+          for (final operation in delta.operations) {
+            if (operation.isInsert && operation.data is Map<String, dynamic>) {
```
Comment (ice-alcides) [inline lib/app/features/feed/create_post/views/pages/create_post_modal/create_post_modal.dart:60]: Could you consider using `records` ([link](https://dart.dev/language/records#multiple-returns)) and `patterns` ([link](https://dart.dev/language/patterns#validating-incoming-json)) to simplify this code?

File: lib/app/features/feed/views/components/text_editor/components/custom_blocks/text_editor_poll_block/components/poll_answers_list_view/poll_answers_list_view.dart [hunk 10]
```diff
@@ -0,0 +1,25 @@
+// SPDX-License-Identifier: ice License 1.0
+
+import 'package:flutter/material.dart';
+import 'package:hooks_riverpod/hooks_riverpod.dart';
+import 'package:ice/app/extensions/num.dart';
+import 'package:ice/app/features/feed/providers/poll/poll_answers_provider.dart';
+import 'package:ice/app/features/feed/views/components/text_editor/components/custom_blocks/text_editor_poll_block/components/poll_answer_item/poll_answer_item.dart';
```
Comment (ice-alcides) [inline lib/app/features/feed/views/components/text_editor/components/custom_blocks/text_editor_poll_block/components/poll_answers_list_view/poll_answers_list_view.dart:7]: I would propose moving `poll_answer_item.dart` to this path:

`import 'package:ice/app/features/feed/views/components/text_editor/components/poll/poll_answer_item.dart';`

File: lib/app/features/feed/views/components/text_editor/components/custom_blocks/text_editor_poll_block/text_editor_poll_block.dart [hunk 11]
```diff
@@ -0,0 +1,83 @@
+// SPDX-License-Identifier: ice License 1.0
+
+import 'package:flutter/material.dart';
+import 'package:flutter_quill/flutter_quill.dart';
+import 'package:ice/app/extensions/extensions.dart';
+import 'package:ice/app/features/feed/views/components/text_editor/components/custom_blocks/text_editor_poll_block/components/poll_add_answer_button/poll_add_answer_button.dart';
```
Comment (ice-alcides) [inline lib/app/features/feed/views/components/text_editor/components/custom_blocks/text_editor_poll_block/text_editor_poll_block.dart:6]:
```
import 'package:ice/app/features/feed/views/components/text_editor/components/custom_blocks/text_editor_poll_block/components/poll_add_answer_button/poll_add_answer_button.dart';
import 'package:ice/app/features/feed/views/components/text_editor/components/custom_blocks/text_editor_poll_block/components/poll_answers_list_view/poll_answers_list_view.dart';
import 'package:ice/app/features/feed/views/components/text_editor/components/custom_blocks/text_editor_poll_block/components/poll_close_button/poll_close_button.dart';
import 'package:ice/app/features/feed/views/components/text_editor/components/custom_blocks/text_editor_poll_block/components/poll_length_button/poll_length_button.dart';
import 'package:ice/app/features/feed/views/components/text_editor/components/custom_blocks/text_editor_poll_block/components/poll_title/poll_title.dart';
```

can you consider decreasing nesting?

File: lib/app/features/feed/views/pages/poll_length_time_modal/poll_length_time_modal.dart [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+                    height: 34.0.s,
+                    margin: EdgeInsets.symmetric(horizontal: 9.0.s),
+                    decoration: BoxDecoration(
+                      color: context.theme.appColors.onSecondaryBackground,
+                      borderRadius: BorderRadius.circular(8.0.s),
+                    ),
+                  ),
+                ),
+              ),
+              Row(
```
Comment (ice-alcides) [inline lib/app/features/feed/views/pages/poll_length_time_modal/poll_length_time_modal.dart:60]: let's extract this `Row` into a standalone class

**Review by ice-kreios — DISMISSED** [review]

File: lib/app/features/feed/create_post/views/pages/create_post_modal/create_post_modal.dart (hunk 9 above)
Comment (ice-alcides) [reply lib/app/features/feed/create_post/views/pages/create_post_modal/create_post_modal.dart]: it's optional, up to you

File: lib/app/features/feed/create_post/views/pages/create_post_modal/create_post_modal.dart (hunk 8 above)
Comment (ice-alcides) [reply lib/app/features/feed/create_post/views/pages/create_post_modal/create_post_modal.dart]: It's used inside callback, so it should be called with `ref.read` or you can get values with `ref.watch`, but then you need to move these lines outside of the `isPollValid` method

**Review by ice-tychon — APPROVED** [review]

**Review by ice-alcides — APPROVED** [review]

**Pull request merged**